Add the quant lift up pass in convert phase

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122777

**Summary**

Lift up the quant node before view like nodes. It can benefit performance of Attention like block. For example, we have the pattern as:

```
          DQ
    DQ       LINEAR
    LINEAR   VIEW
    VIEW     PERMUTE
    PERMUTE  TRANSPOSE
    Q        Q
    DQ       DQ
       Matmul
        DIV
        ADD
      SOFTMAX
```

We want to lift up the the quant nodes from `matmul` before view like nodes as the output of Linear node.

```
          DQ
    DQ       LINEAR
    LINEAR   Q
    Q        VIEW
    VIEW     PERMUTE
    PERMUTE  TRANSPOSE
    DQ       DQ
       Matmul
        DIV
        ADD
      SOFTMAX
```

It produces a `DQ->LINEAR->Q` pattern which can be fused by backend.

**Test Plan**

```
python -m pytest quantization/pt2e/test_x86inductor_quantizer.py -k test_attention_block
```


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

leslie-fang-intel committed:
Add the quant lift up pass in convert phase

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/pt2e/quant_lift_up_pass.py:
```diff
@@ -0,0 +1,93 @@
+]
+
+
+class QuantLiftUp(PassBase):
```
we discussed this before, so this is one way to do it, the other way is just to annotate view, permute, transpose with SharedQuantizationSpec between output and input, and we'll fuse the (dq -> view -> q) pattern in lowering to be a single view op that works with quantized tensor. And last time I think you preferred this approach. I also feel observing these ops will be better since you'll be able to simulate quantization the same way the other ops are doing, also good for QAT as well. Is there a reason to do this lifting instead of the observation approach?

leslie-fang-intel replied:
Thanks for the comment. Yean, I have another discussion with @chunyuan-w, she mentions a case for the downside of this `observation approach`. For example a pattern as
```
linear
view
fp32_op doesn't support int8
```
1. The lift up pass approach will generate pattern as 
```
dq
linear
view
fp32_op
```
2. The observation approach will generate pattern as
```
dq
linear
q
dq
view
fp32_op
```
In this case, the observation approach will do redundant q/dq after the linear op.

Also cc @jgong5 for this discussion.

jgong5 replied:
Ah yes, seems like forcing the quantization for the ops like view, transpose, permute would introduce side-effect on accuracy and performance like the case @leslie-fang-intel pointed out. @jerryzh168 any thoughts on this?

jerryzh168 replied:
I believe these two passes are equivalent in terms of how much quantization patterns or ways of quantization they can express, the only difference is 2 is more explicit than 1.

in this specific case, what you can do is to decide whether to observe view or not based on the annotation of previous operator, if the previous operator (linear) is observed, then we'll observe it as well, otherwise we'll just skip annotation. this is more explicit than the lifting pass I think.

leslie-fang-intel replied:
> in this specific case, what you can do is to decide whether to observe view or not based on the annotation of previous operator, if the previous operator (linear) is observed, then we'll observe it as well, otherwise we'll just skip annotation. 

Thanks and I think that's exactly the same recipe of `view` we mentioned above in https://github.com/pytorch/pytorch/pull/122777#discussion_r1543944541 with the `observation approach`. In which, it brings a extra pair of `Quant/Dequant` at the output of `linear` comparing to `lift up pass` for this specific case.

To avoid it in `observation approach`, we may have to check the following up node of `view` whether it can be quantized (beside input node is observed or not). Then we will decide to observer `view` or not. Any thoughts?

jerryzh168 replied:
I see, so to summarize, whether the output of linear should be observed or not depends on whether view is followed by a quantizable op?

jerryzh168 replied:
this makes sense I think, in that case it's probably better to do in lowering to not complicate the logic in annotation. but I'm wondering if you could do this in intel lowering code base, instead of in quantization code base?

leslie-fang-intel replied:
> I see, so to summarize, whether the output of linear should be observed or not depends on whether view is followed by a quantizable op?

Thanks for the comment. In current recipe, if linear followed by a quantizable op, we will annotate the input of this quantizable operator (which is also the output of this linear). And for view, the logic in annotation will be complicate.

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Add the quant lift up pass in convert phase"


**Summary**

Lift up the quant node before view like nodes. It can benefit performance of Attention like block. For example, we have the pattern as:

```
          DQ
    DQ       LINEAR
    LINEAR   VIEW
    VIEW     PERMUTE
    PERMUTE  TRANSPOSE
    Q        Q
    DQ       DQ
       Matmul
        DIV
        ADD
      SOFTMAX
```

We want to lift up the the quant nodes from `matmul` before view like nodes as the output of Linear node.

```
          DQ
    DQ       LINEAR
    LINEAR   Q
    Q        VIEW
    VIEW     PERMUTE
    PERMUTE  TRANSPOSE
    DQ       DQ
       Matmul
        DIV
        ADD
      SOFTMAX
```

It produces a `DQ->LINEAR->Q` pattern which can be fused by backend.

**Test Plan**

```
python -m pytest quantization/pt2e/test_x86inductor_quantizer.py -k test_attention_block
```


[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Add the quant lift up pass in convert phase"


**Summary**

Lift up the quant node before view like nodes. It can benefit performance of Attention like block. For example, we have the pattern as:

```
          DQ
    DQ       LINEAR
    LINEAR   VIEW
    VIEW     PERMUTE
    PERMUTE  TRANSPOSE
    Q        Q
    DQ       DQ
       Matmul
        DIV
        ADD
      SOFTMAX
```

We want to lift up the the quant nodes from `matmul` before view like nodes as the output of Linear node.

```
          DQ
    DQ       LINEAR
    LINEAR   Q
    Q        VIEW
    VIEW     PERMUTE
    PERMUTE  TRANSPOSE
    DQ       DQ
       Matmul
        DIV
        ADD
      SOFTMAX
```

It produces a `DQ->LINEAR->Q` pattern which can be fused by backend.

**Test Plan**

```
python -m pytest quantization/pt2e/test_x86inductor_quantizer.py -k test_attention_block
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

jerryzh168 reviewed (approved):
looks good to me, thanks for addressing the comments!

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

jgong5 commented on torch/_inductor/fx_passes/quantization.py:
```diff
@@ -2163,3 +2176,78 @@ def _register_quantization_weight_pack_pass():
+        if (
+            node.op == "call_function"
+            and node.target in _QUANTIZE_PER_TENSOR_OPS
+            and len(node.all_input_nodes) == 1
```
Why is this check needed?

leslie-fang-intel replied:
As synced offline and add the comments for this check.

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "Add the quant lift up pass in convert phase"


**Summary**

Lift up the quant node before view like nodes. It can benefit performance of Attention like block. For example, we have the pattern as:

```
          DQ
    DQ       LINEAR
    LINEAR   VIEW
    VIEW     PERMUTE
    PERMUTE  TRANSPOSE
    Q        Q
    DQ       DQ
       Matmul
        DIV
        ADD
      SOFTMAX
```

We want to lift up the the quant nodes from `matmul` before view like nodes as the output of Linear node.

```
          DQ
    DQ       LINEAR
    LINEAR   Q
    Q        VIEW
    VIEW     PERMUTE
    PERMUTE  TRANSPOSE
    DQ       DQ
       Matmul
        DIV
        ADD
      SOFTMAX
```

It produces a `DQ->LINEAR->Q` pattern which can be fused by backend.

**Test Plan**

```
python -m pytest quantization/pt2e/test_x86inductor_quantizer.py -k test_attention_block
```


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

leslie-fang-intel committed:
Update

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
